### DISCUSSION TITLE
Endpoints - support form submissions for noscript and fetch users alike

### DISCUSSION THREAD

**Discussion by tmaxmax** [sveltejs/kit#3326] [Q&A] [answered]
I am developing a website that must support users without Javascript - this means that I must ensure HTML forms work properly without any enhancements. This raises a problem: all my endpoints are designed as part of a REST API, so they return JSON, but when working with basic HTML forms only, the server must handle redirections and respond with actual HTML.
How would you approach this? Do I need to make two endpoints for the same resource? Or am I misunderstanding something, and this is not a real problem? Thank you for your help!

**Comment by maxicarlos08** [comment]
What I did was add an additional `source=form` query parameter to the form action, so I could return a redirect redirect to some results page when using a non-javascript form

**Comment by geoffrich** [comment, marked as answer]
The SvelteKit starter app does this via an [`enhance` action](https://github.com/sveltejs/kit/blob/master/packages/create-svelte/templates/default/src/lib/form.ts) on the form. If JS is enabled, then the `enhance` action runs and submits the form via fetch. Otherwise, it does a regular form submit. 

In the `/form` endpoint, it detects if the submission was via fetch or regular form submission. If it was a regular form submission, then it [performs a redirect](https://github.com/sveltejs/kit/blob/f17143414cf2efd966c3432bc29e6133908e75f2/packages/create-svelte/templates/default/src/routes/todos/_api.ts#L35-L39) instead of returning JSON.

You can see further discussion on this issue: https://github.com/sveltejs/kit/issues/1203#issuecomment-826118551.
